### PR TITLE
Issue 5800: Optimize the ManifestBuilder

### LINKED ISSUE #5800: Optimize the ManifestBuilder
The content builder takes an inordinate amount of time.
The main culprit is the `ManifestBuilder.getDependants()` which does things in an O(N^2) time complexity.

`game.project`:
* GameProjectBuilder.createArchive took 77.316000
* ManifestBuilder.buildManifestData took 344.338000
* Task game.project took 427.038000 s

All content tasks took 471.820000 s (including game.project)

Entire build took 511.35s (not including engine build)

### PR DESCRIPTION
**Pull request by JCash**

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java [hunk 1]
```diff
@@ -1167,6 +1179,7 @@ private void registerPipelinePlugins() {
                 boolean abort = false;
                 try {
                     builder.build(task);
+                    monitor.worked(1);
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java:1182]: DoN't count up until the work is done.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/archive/ManifestBuilder.java [hunk 2]
```diff
@@ -454,35 +458,29 @@ public void addResourceEntry(String url, byte[] data, int flags) throws IOExcept
            and thus create a partial archive that has to be updated (through
            LiveUpdate) before that CollectionProxy can be loaded.
         */
-        ResourceNode candidate = null;
+
+        List<String> dependants = new ArrayList<String>();
+
+        ResourceNode candidate = pathToNode.get(filepath);
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/archive/ManifestBuilder.java:464]: Don't traverse the entire tree to find the node.
Saved `68` seconds(!) in the test project.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/archive/ManifestBuilder.java [hunk 3]
```diff
@@ -504,16 +502,28 @@ public ManifestHeader buildManifestHeader() throws IOException {
         return builder.build();
     }
 
+    private void buildPathToNodeMap(ResourceNode node) {
+        pathToNode.put(node.relativeFilepath, node);
+        for (ResourceNode child : node.getChildren()) {
+            buildPathToNodeMap(child);
+        }
+    }
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/archive/ManifestBuilder.java:510]: Build the path-to-node map once.

**Comment by JCash** [comment]
With these changes, the build time for the manifest data went from `420 seconds` to `0.1 seconds`

**Comment by britzl** [comment]
That's a very impressive speedup! Well done.

**Review by Jhonnyg — DISMISSED** [review]
Looks good! 👍

**Comment by Jhonnyg** [comment]
I agree with Björn, amazing results!

**Pull request merged**